File: pkgs/objective_c/src/objective_c.c
// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#include "objective_c.h"

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <dispatch/dispatch.h>

#include "include/dart_api_dl.h"
#include "objective_c_runtime.h"

_Atomic bool _mainThreadIsListening = false;

static void setListening(void* arg) {
  _mainThreadIsListening = true;
}

FFI_EXPORT intptr_t DOBJC_initializeApi(void* data) {
  dispatch_async_f(dispatch_get_main_queue(), NULL, setListening);
  return Dart_InitializeApiDL(data);
}

// Copy and dispose helper for ObjC blocks that wrap a Dart closure. For these
// blocks, the target is an int ID, and the dispose_port is listening for these
// IDs.
FFI_EXPORT void DOBJC_copyObjCBlockWithClosure(ObjCBlockImpl* dst,
                                               ObjCBlockImpl* src) {
  dst->target = src->target;
  dst->dispose_port = src->dispose_port;
}

FFI_EXPORT void DOBJC_disposeObjCBlockWithClosure(ObjCBlockImpl* block) {
  Dart_PostInteger_DL(block->dispose_port, (int64_t)block->target);
}

FFI_EXPORT bool DOBJC_isValidBlock(ObjCBlockImpl* block) {
  if (block == NULL) return false;
  void* isa = block->isa;
  return isa == &_NSConcreteStackBlock || isa == &_NSConcreteMallocBlock ||
         isa == &_NSConcreteAutoBlock || isa == &_NSConcreteFinalizingBlock ||
         isa == &_NSConcreteGlobalBlock || isa == &_NSConcreteWeakBlockVariable;
}

FFI_EXPORT void DOBJC_finalizeObject(void* isolate_callback_data, void* peer) {
  // objc_release works for Objects and Blocks.
  DOBJC_runOnMainThread((void (*)(void*))objc_release, peer);
}

FFI_EXPORT Dart_FinalizableHandle
DOBJC_newFinalizableHandle(Dart_Handle owner, ObjCObjectImpl* object) {
  return Dart_NewFinalizableHandle_DL(owner, object, 0, DOBJC_finalizeObject);
}

FFI_EXPORT void DOBJC_deleteFinalizableHandle(Dart_FinalizableHandle handle,
                                              Dart_Handle owner) {
  Dart_DeleteFinalizableHandle_DL(handle, owner);
}

static void finalizeMalloc(void* isolate_callback_data, void* peer) {
  free(peer);
}

FFI_EXPORT bool* DOBJC_newFinalizableBool(Dart_Handle owner) {
  bool* pointer = (bool*)malloc(1);
  *pointer = false;
  Dart_NewFinalizableHandle_DL(owner, pointer, 1, finalizeMalloc);
  return pointer;
}

FFI_EXPORT bool DOBJC_postCObject(int64_t port_id, void* peer,
                                  void (*callback)(void*, void*)) {
  Dart_CObject c_post;
  c_post.type = Dart_CObject_kNativePointer;
  c_post.value.as_native_pointer.ptr = (intptr_t)peer;
  c_post.value.as_native_pointer.size = 0;
  c_post.value.as_native_pointer.callback = (Dart_HandleFinalizer)callback;
  if (!Dart_PostCObject_DL(port_id, &c_post)) {
    if (c_post.value.as_native_pointer.callback != NULL) {
      c_post.value.as_native_pointer.callback(
          NULL, (void*)c_post.value.as_native_pointer.ptr);
    }
    return false;
  }
  return true;
}

FFI_EXPORT DOBJC_Context* DOBJC_fillContext(DOBJC_Context* context) {
  context->version = 2;
  context->newWaiter = DOBJC_newWaiter;
  context->awaitWaiter = DOBJC_awaitWaiter;
  context->currentIsolate = Dart_CurrentIsolate_DL;
  context->enterIsolate = Dart_EnterIsolate_DL;
  context->exitIsolate = Dart_ExitIsolate_DL;
  context->getMainPortId = Dart_GetMainPortId_DL;
  context->getCurrentThreadOwnsIsolate = Dart_GetCurrentThreadOwnsIsolate_DL;
  context->postCObject = DOBJC_postCObject;
  return context;
}

File: pkgs/objective_c/src/objective_c.m
// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#import <Foundation/Foundation.h>
#import <Foundation/NSDate.h>
#import <Foundation/NSThread.h>
#import <dispatch/dispatch.h>
#import <objc/message.h>

#if !__has_feature(objc_arc)
#error "This file must be compiled with ARC enabled"
#endif

#include "ffi.h"
#include "include/dart_api_dl.h"
#include "os_version.h"

_Atomic bool _mainThreadIsListening = false;

FFI_EXPORT intptr_t DOBJC_initializeApi(void* data) {
  dispatch_async(dispatch_get_main_queue(), ^{
    _mainThreadIsListening = true;
  });
  return Dart_InitializeApiDL(data);
}

FFI_EXPORT void DOBJC_runOnMainThread(void (*fn)(void *), void *arg) {
  if (!_mainThreadIsListening || [NSThread isMainThread]) {
    fn(arg);
  } else {
    dispatch_async(dispatch_get_main_queue(), ^{
      fn(arg);
    });
  }
}

@interface DOBJCWaiter : NSObject {}
@property(strong) NSCondition* cond;
@property bool done;
-(void)signal;
-(void)wait;
@end

@implementation DOBJCWaiter
-(instancetype)init {
  self = [super init];
  if (self) {
    _cond = [[NSCondition alloc] init];
    _done = false;
  }
  return self;
}
-(void)signal {
  [_cond lock];
  _done = true;
  [_cond signal];
  [_cond unlock];
}
-(void)wait {
  [_cond lock];
  while (!_done) {
    [_cond wait];
  }
  [_cond unlock];
}
@end

FFI_EXPORT void* DOBJC_newWaiter(void) {
  DOBJCWaiter* w = [[DOBJCWaiter alloc] init];
  // __bridge_retained increments the ref count, __bridge_transfer decrements
  // it, and __bridge doesn't change it. One of the __bridge_retained calls is
  // balanced by the __bridge_transfer in signalWaiter, and the other is
  // balanced by the one in awaitWaiter. In other words, this function returns
  // an object with a +2 ref count, and signal and await each decrement the
  // ref count.
  return (__bridge_retained void*)(__bridge id)(__bridge_retained void*)(w);
}

FFI_EXPORT void DOBJC_signalWaiter(void* waiter) {
  if (waiter) [(__bridge_transfer DOBJCWaiter*)waiter signal];
}

FFI_EXPORT void DOBJC_awaitWaiter(void* waiter) {
  [(__bridge_transfer DOBJCWaiter*)waiter wait];
}

FFI_EXPORT Version DOBJC_getOsVesion(void) {
  NSOperatingSystemVersion objc_version =
      [[NSProcessInfo processInfo] operatingSystemVersion];
  Version c_version;
  c_version.major = (int)objc_version.majorVersion;
  c_version.minor = (int)objc_version.minorVersion;
  c_version.patch = (int)objc_version.patchVersion;
  return c_version;
}

void DOBJC_freePortBlockArgs(void* peer) {
  @autoreleasepool {
    id args = (__bridge_transfer id)peer;
    (void)args;
  }
}

void DOBJC_finalizePortBlockArgs(void* _, void* peer) {
  DOBJC_runOnMainThread(DOBJC_freePortBlockArgs, peer);
}

void DOBJC_finalizePortBlockWaiter(void* _, void* peer) {
  DOBJC_signalWaiter(peer);
}

FFI_EXPORT void DOBJC_invokeListenerPortBlock(int64_t port, void* args) {
  Dart_CObject cobj;
  cobj.type = Dart_CObject_kNativePointer;
  cobj.value.as_native_pointer.ptr = (intptr_t)args;
  cobj.value.as_native_pointer.size = 0;
  cobj.value.as_native_pointer.callback = DOBJC_finalizePortBlockArgs;

  if (!Dart_PostCObject_DL(port, &cobj)) {
    DOBJC_finalizePortBlockArgs(NULL, args);
  }
}

FFI_EXPORT void DOBJC_invokeBlockingPortBlock(
    int64_t port, void* args, void* waiter) {
  Dart_CObject cargs;
  cargs.type = Dart_CObject_kNativePointer;
  cargs.value.as_native_pointer.ptr = (intptr_t)args;
  cargs.value.as_native_pointer.size = 0;
  cargs.value.as_native_pointer.callback = DOBJC_finalizePortBlockArgs;

  Dart_CObject cwaiter;
  cwaiter.type = Dart_CObject_kNativePointer;
  cwaiter.value.as_native_pointer.ptr = (intptr_t)waiter;
  cwaiter.value.as_native_pointer.size = 0;
  cwaiter.value.as_native_pointer.callback = DOBJC_finalizePortBlockWaiter;

  Dart_CObject* cobjArray[] = {&cargs, &cwaiter};
  Dart_CObject cobj;
  cobj.type = Dart_CObject_kArray;
  cobj.value.as_array.values = cobjArray;
  cobj.value.as_array.length = sizeof(cobjArray) / sizeof(cobjArray[0]);

  if (!Dart_PostCObject_DL(port, &cobj)) {
    DOBJC_finalizePortBlockArgs(NULL, args);
    DOBJC_finalizePortBlockWaiter(NULL, waiter);
  }
}

@interface PortBlockFinalizer : NSObject {
  int64_t port;
}
- (instancetype)initWithPort:(int64_t) _port;
@end

@implementation PortBlockFinalizer
- (instancetype)initWithPort:(int64_t) _port {
  if (self = [super init]) {
    port = _port;
  }
  return self;
}
- (void)dealloc {
  Dart_CObject cobj;
  cobj.type = Dart_CObject_kNull;
  Dart_PostCObject_DL(port, &cobj);
}
@end

static const char BLOCK_FINALIZER_KEY;

FFI_EXPORT void DOBJC_attachPortBlockFinalizer(void* block, int64_t port) {
  PortBlockFinalizer* dtor = [[PortBlockFinalizer alloc] initWithPort:port];
  objc_setAssociatedObject(
      (__bridge id)block, &BLOCK_FINALIZER_KEY, dtor,
      OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}
